setHoldTime.js

```diff
@@ -0,0 +1,131 @@
+const Twilio = require('twilio');
+const { sleep } = require('./utils');
+
+const setHoldTime = async (client, reservationSid, taskSid, timestamp, retryCount) => {
+  let currentRetry = retryCount ? retryCount : 0;
+
+  console.log('Checking hold time for reservation', reservationSid);
+
+  try {
+    // Check for hold time sync doc for this reservation
+    // First, generate an access token for Sync
+    const AccessToken = Twilio.jwt.AccessToken;
+    const SyncGrant = AccessToken.SyncGrant;
+    
+    const token = new AccessToken(
+      ACCOUNT_SID,
+      API_KEY,
+      API_SECRET
+    );
+    
+    token.identity = 'TaskWrapService';
+    
+    const syncGrant = new SyncGrant({
+      serviceSid: SYNC_SERVICE_SID || 'default'
+    });
+    
+    token.addGrant(syncGrant);
+    const tokenStr = token.toJwt();
+    
+    const syncClient = new SyncClient(tokenStr);
+    const docName = `${reservationSid}_HoldTime`;
+    let holdTime = 0;
+    
+    try {
+      // Attempt to open the Sync doc
+      let doc = await syncClient.document({ id: docName, mode: 'open_existing' });
+      let { data } = doc;
+      
+      if (!data.holdTime) {
+        // no holds
+        holdTime = 0;
+      } else {
+        holdTime = data.holdTime;
+      }
+      
+      if (!isNaN(data.currentHoldStart) && data.currentHoldStart > 0) {
+        // hold in progress, add to existing hold time
+        const currentHoldDuration = (timestamp - data.currentHoldStart) / 1000;
+        
+        holdTime = holdTime + currentHoldDuration;
+        
+        console.log(`Adding ${currentHoldDuration} seconds of outstanding hold time found for reservation`, reservationSid);
+      }
+      
+    } catch (error) {
+      if (error.status === 404) {
+        // doc didn't exist
+        console.log(`No hold time doc found for reservation`, reservationSid);
+      } else {
+        console.error(`Error getting hold time doc for reservation ${reservationSid}`, error);
+      }
+    }
+    
+    // get attributes for the task, update if hold time differs
+    console.log('Checking conversations.hold_time task attribute for reservation', reservationSid);
+    const task = await client.taskrouter
+      .workspaces(TASKROUTER_WORKSPACE_SID)
+      .tasks(taskSid)
+      .fetch();
+    
+    let { attributes: attributesStr } = task;
+    let attributes = {};
+    
+    if (attributesStr) {
+      attributes = JSON.parse(attributesStr);
+    }
+    
+    if (!attributes.conversations || isNaN(attributes.conversations.hold_time) || attributes.conversations.hold_time !== holdTime) {
+      console.log(`Updating conversations.hold_time task attribute to ${holdTime} seconds on task ${taskSid} from reservation`, reservationSid);
+      
+      attributes = {
+        ...attributes,
+        conversations: {
+          ...attributes.conversations,
+          hold_time: holdTime
+        }
+      };
+      
+      await client.taskrouter
```

Comment:
Since there's a risk we could be attempting to update the task attributes as the same time as another process, like another Flex UI instance handling a different reservation for this same task, we should leverage TaskRouter's conflict resolution features.

https://www.twilio.com/docs/usage/mutation-and-conflict-resolution